Add :IMPORT-FROM option to DEFPACKAGE

MattHeffron opened the pull request:
This resolves #2334

pamoroso commented:
When building this PR on Linux Mint 22.1 Cinnamon I get a blank Medley window and the following error:

```
paolo@lispmachine:~/medley/medley$ ./scripts/loadup-all.sh -apps
>>>>> START loadup-init
"/home/paolo/bin/lde" "/home/paolo/medley/medley/internal/loadups/starter.sysout" -id "loadup_init_1" -title "Medley::loadup_init_1" -g 1024x768 -sc 1024x768 -noscroll
MEDLEYDIR: "/home/paolo/medley/medley"
LOGINDIR: "/home/paolo/medley/medley/loadups/build/logindir"
GREET FILE: "/home/paolo/medley/medley/loadups/build/loadup-init.init"
REM.CM FILE: ""
VMEM FILE: "/home/paolo/medley/medley/loadups/build/logindir/vmem/lisp_loadup_init_1.virtualmem"
+++++ SUCCESS +++++
..... files created .....
-rw-rw-r-- 1 paolo paolo 4426752 Oct 30 11:03 /home/paolo/medley/medley/loadups/build/init.dlinit
-rw-rw-r-- 1 paolo paolo 150218 Oct 30 11:03 /home/paolo/medley/medley/loadups/build/init.dribble
-rw-rw-r-- 1 paolo paolo 4279808 Oct 30 11:03 /home/paolo/medley/medley/loadups/build/init.sysout
-rw-rw-r-- 1 paolo paolo 55826 Oct 30 11:02 /home/paolo/medley/medley/loadups/build/RDSYS
-rw-rw-r-- 1 paolo paolo 43012 Oct 30 11:02 /home/paolo/medley/medley/loadups/build/RDSYS.LCOM
-rw-rw-r-- 1 paolo paolo 89183 Oct 30 11:02 /home/paolo/medley/medley/loadups/build/I-NEW
-rw-rw-r-- 1 paolo paolo 70676 Oct 30 11:02 /home/paolo/medley/medley/loadups/build/I-NEW.LCOM
<<<<< END loadup-init

>>>>> START loadup-mid-from-init
"/home/paolo/bin/ldeinit" "/home/paolo/medley/medley/loadups/build/init.dlinit" -id "loadup_mid_from_init_1" -title "Medley::loadup_mid_from_init_1" -g 1024x768 -sc 1024x768 -noscroll -NF
MEDLEYDIR: "/home/paolo/medley/medley"
LOGINDIR: "/home/paolo/medley/medley/loadups/build/logindir"
GREET FILE: "/home/paolo/medley/medley/greetfiles/NOGREET"
REM.CM FILE: "/home/paolo/medley/medley/loadups/build/loadup-mid-from-init.cm"
VMEM FILE: "/home/paolo/medley/medley/loadups/build/init-mid.sysout"
evaluating initial expressions:
PACKAGE-CONVERSION-TABLE.EXPRESSIONS
LLSUBRS.EXPRESSIONS
FILEIO.EXPRESSIONS
EXTERNALFORMAT.EXPRESSIONS

*Error* URAID Called:
Enter the URaid
#:CL::SYMBOL-PACKAGE

< 
```

<img width="1024" height="800" alt="pr2335error" src="https://github.com/user-attachments/assets/87ddbbed-326c-48af-8cbd-caf3bcbf36b1" />

I'm at mth52--Add-IMPORT-FROM-to-DEFPACKAGE/c2d89e5 on Medley and nhb-rewrite-version-parser-v2/3d09555 on Maiko.

rmkaplan commented:
Doesn't it have to be FAKE-COMPILE-FILE?  It has a lot of commonlisp FUNCTIONS etc.

> On Oct 30, 2025, at 11:33 AM, Nick Briggs ***@***.***> wrote:
> 
> 
> nbriggs
>  left a comment 
> (Interlisp/medley#2335)
>  <https://github.com/Interlisp/medley/pull/2335#issuecomment-3469498167>
> The previous LCOM was made with compile-file, according to the comments in it, as was the version in CLTL2. So, i'm not sure if that's COMPILE-FILE or FAKE-COMPILE-FILE... since it's an LCOM.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/Interlisp/medley/pull/2335#issuecomment-3469498167>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AQSTUJP2ZIXP6T4Q7RAHD5D32JKX7AVCNFSM6AAAAACKUGUZBSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTINRZGQ4TQMJWG4>.
> You are receiving this because you are subscribed to this thread.
>

nbriggs commented:
I can do a loadup with the revised LLPACKAGE.LCOM.  I don't have anything to test the :IMPORT-FROM functionality.

pamoroso commented:
It now builds successfully but I'm not sure I'm testing the feature properly as I'd expect the call to `CREATEW` to work here:

<img width="503" height="350" alt="defpackage-error" src="https://github.com/user-attachments/assets/a36a60b5-b2a6-418f-b115-9225c3bbf21d" />

masinter reviewed (approved):
I don't think this PR needs to wait for it, but we should try to make the apperance of internal package prefix notation in our pretty printed source code. It's especially egregious to have :: double colon references to symbols which are internal lexically scoped variables. 

I'm not sure where to put the fix to eliminate those uses.

MattHeffron commented:
> the existing DEFPACKAGE code seems to have carefully used IL: symbols for all the temporary variables

If it is thought to be appropriate, I can redo this using `IL:` package local variables.

pamoroso commented:
What am I doing wrong in this branch?

<img width="541" height="312" alt="pkg-err" src="https://github.com/user-attachments/assets/c7cc1e28-d434-4300-ae9b-990cabb632d4" />

masinter reviewed (approved):
IMPORT PACKAGE param fixed

MattHeffron commented:
Fixed error of missing PACKAGE argument in call to IMPORT.
Rebuilt loadup successfully.
Tested by:
1. In XCL exec, verifying no symbol "IT" in the "XCL-USER" package. `(find-symbol "IT")` returned `NIL`
2. Change to "USER" package in the exec.
3. Verifying no symbol "IT" in the "USER" package. `(find-symbol "IT")` returned `NIL`
4. Gave command in the same exec:
```common-lisp
(IL:DEFPACKAGE "XCL-USER" (:IMPORT-FROM "IL" "IT"))
``` 
5. Changed back to "XCL-USER" package in the exec.
6. Verify that symbol "IT" is accessible from that package, and that it is the same symbol as `IL:IT`
I.e., `(find-symbol "IT")` returns 2 values:
`IT`
`:INTERNAL`
7. Verify that it is the _same_ symbol as `IL:IT`
`(EQ 'IT 'IL:IT)` returns `T`
